builtins/web/blob.cpp

```diff
@@ -0,0 +1,703 @@
+#include "js/Stream.h"
+#include "js/TypeDecls.h"
+#include "js/Value.h"
+
+#include <regex>
+
+namespace {
+
+static api::Engine *ENGINE;
+
+JSObject* new_array_buffer_from_span(JSContext *cx, std::span<const uint8_t> span) {
+  auto buf = mozilla::MakeUnique<uint8_t[]>(span.size());
+  if (!buf) {
+    JS_ReportOutOfMemory(cx);
+    return nullptr;
+  }
+
+  std::copy(span.begin(), span.end(), buf.get());
+
+  auto array_buffer = JS::NewArrayBufferWithContents(
+      cx, span.size(), buf.get(), JS::NewArrayBufferOutOfMemory::CallerMustFreeMemory);
+  if (!array_buffer) {
+    JS_ReportOutOfMemory(cx);
+    return nullptr;
+  }
+
+  // `array_buffer` now owns `buf`
+  std::ignore = (buf.release());
+  return array_buffer;
+}
+
+template <typename T> bool validate_type(T *chars, size_t strlen) {
+  for (size_t i = 0; i < strlen; i++) {
+    T c = chars[i];
+    if (c < 0x20 || c > 0x7E) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+// 1. If type contains any characters outside the range U+0020 to U+007E, then set t to the empty string.
+// 2. Convert every character in type to ASCII lowercase.
+JSString *normalize_type(JSContext *cx, HandleValue value) {
+  JS::RootedString value_str(cx);
+
+  if (value.isObject() || value.isString()) {
+    value_str = JS::ToString(cx, value);
+    if (!value_str) {
+      return nullptr;
+    }
+  } else if (value.isNull()) {
+    return JS::ToString(cx, value);
+  } else {
+    return JS_GetEmptyString(cx);
+  }
+
+  auto str = JS::StringToLinearString(cx, value_str);
+  if (!str) {
+    return nullptr;
+  }
+
+  std::string normalized;
+  auto strlen = JS::GetLinearStringLength(str);
+
+  if (!strlen) {
+    return JS_GetEmptyString(cx);
+  }
+
+  if (JS::LinearStringHasLatin1Chars(str)) {
+    JS::AutoCheckCannotGC nogc(cx);
+    auto chars = JS::GetLatin1LinearStringChars(nogc, str);
+    if (!validate_type(chars, strlen)) {
+      return JS_GetEmptyString(cx);
+    }
+
+    normalized = std::string(reinterpret_cast<const char *>(chars), strlen);
+  } else {
+    JS::AutoCheckCannotGC nogc(cx);
+    auto chars = (JS::GetTwoByteLinearStringChars(nogc, str));
+    if (!validate_type(chars, strlen)) {
+      return JS_GetEmptyString(cx);
+    }
+
+    normalized.reserve(strlen);
+    for (size_t i = 0; i < strlen; ++i) {
+      normalized += static_cast<unsigned char>(chars[i]);
+    }
+  }
+
+  std::transform(normalized.begin(), normalized.end(), normalized.begin(),
+                 [](unsigned char c) { return std::tolower(c); });
+
+  return JS_NewStringCopyN(cx, normalized.c_str(), normalized.length());
+}
+
+// https://w3c.github.io/FileAPI/#convert-line-endings-to-native
+std::string convert_line_endings_to_native(std::string_view s) {
+  std::string native_line_ending = "\n";
+#ifdef _WIN32
+  native_line_ending = "\r\n";
+#endif
+
+  std::string result;
+  std::regex re(R"(\r\n|\r|\n)");
+  std::regex_replace(std::back_inserter(result), s.begin(), s.end(), re, native_line_ending);
```

Comment:
I'd prefer not pulling in `std::regex` for this, because it's pretty heavy. Could you instead do this as a manual loop? Basically, find each instance of `\r`, check if the next char is `\n`, then either replace or remove `\r` accordingly?

(This of course only works for non-Windows. Currently, that doesn't matter, since we only compile for Wasm. If you want to prepare things for a possible future in which we have native builds, you could add an `#ifdef` implementing the logic for that—up to you.)